File: python/tvm/topi/vision/max_pool2d_location.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# pylint: disable=invalid-name, too-many-nested-blocks
"""max_pool2d_location operator"""
from ... import te, tir


def _max_pool2d_location(input_data, out_buf, kernel_h, kernel_w, stride_h, stride_w, pad_t, pad_l):

    ib = tir.ir_builder.create()
    data = ib.buffer_ptr(input_data)
    out = ib.buffer_ptr(out_buf)
    temp = ib.allocate("float32", (1,), name="temp_data", scope="local")

    numb_, channels_, height_, width_ = input_data.shape
    pooled_height_ = out_buf.shape[2]
    pooled_width_ = out_buf.shape[3]

    with ib.for_range(0, numb_) as n:
        with ib.for_range(0, channels_) as c:
            with ib.for_range(0, pooled_height_) as ph:
                with ib.for_range(0, pooled_width_) as pw:
                    hstart = ph * stride_h - pad_t
                    wstart = pw * stride_w - pad_l
                    hend = te.min(hstart + kernel_h, height_)
                    wend = te.min(wstart + kernel_w, width_)
                    hstart = te.max(hstart, 0)
                    wstart = te.max(wstart, 0)
                    i_index = (
                        n * channels_ * height_ * width_
                        + c * height_ * width_
                        + hstart * width_
                        + wstart
                    )
                    o_index = (
                        n * channels_ * pooled_height_ * pooled_width_
                        + c * pooled_height_ * pooled_width_
                        + ph * pooled_width_
                        + pw
                    )
                    temp[0] = data[i_index]
                    out[o_index] = tir.Cast("float32", hstart * width_ + wstart)

                    with ib.for_range(0, hend) as h:
                        with ib.if_scope(h >= hstart):
                            with ib.for_range(0, wend) as w:
                                with ib.if_scope(w >= wstart):
                                    t_index = (
                                        n * channels_ * height_ * width_
                                        + c * height_ * width_
                                        + h * width_
                                        + w
                                    )
                                    with ib.if_scope(data[t_index] > temp[0]):
                                        temp[0] = data[t_index]
                                        out[o_index] = tir.Cast("float32", h * width_ + w)

    return ib.get()


def max_pool2d_location(
    data, pool_size=(1, 1), strides=(1, 1), padding=(0, 0), layout="NCHW", ceil_mode=False
):
    """Max pool2d location op

    Parameters
    ----------
    data : tvm.Tensor
        4-D with shape [batch, channel, height, width]

    pool_size : int or tuple of int, optional
        The size of window for pooling.

    strides : tuple of int, optional
        The strides of pooling.

    padding : tuple of int, optional
        The padding for pooling.

    ceil_mode : bool, optional
        To enable or disable ceil while pooling.

    Returns
    -------
    result : tvm.relay.Expr
        The computed result.
    """
    assert layout == "NCHW", "Layout of max_pool2d_location must be NCHW"
    assert ceil_mode, "Ceil_mode must be true."
    kernel_h, kernel_w = pool_size
    stride_h, stride_w = strides
    pad_t, pad_l, pad_d, pad_r = padding
    pad_h = pad_t + pad_d
    pad_w = pad_l + pad_r

    numb_, channels_, height_, width_ = data.shape
    tmp_h = height_ + pad_h - kernel_h
    tmp_h_ = te.div(tmp_h, stride_h)
    tmp_w = width_ + pad_w - kernel_w
    tmp_w_ = te.div(tmp_w, stride_w)
    pooled_height_ = te.ceil(tmp_h_) + 1
    pooled_width_ = te.ceil(tmp_w_) + 1

    top_mask = te.extern(
        [numb_, channels_, pooled_height_, pooled_width_],
        [data],
        lambda ins, outs: _max_pool2d_location(
            ins[0], outs[0], kernel_h, kernel_w, stride_h, stride_w, pad_t, pad_l
        ),
    )

    return top_mask

File: src/relay/op/vision/max_pool2d_location.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file max_pool2d_location.cc
 * \brief max_pool2d_location operators
 */
#include <tvm/relay/attrs/vision.h>
#include <tvm/relay/op.h>
#include <tvm/relay/op_attr_types.h>

namespace tvm {
namespace relay {

TVM_REGISTER_NODE_TYPE(MaxPool2dLocationAttrs);

bool MaxPool2dLocationRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
                          const TypeReporter& reporter) {
  CHECK_EQ(types.size(), 2);
  const auto* data = types[0].as<TensorTypeNode>();

  if (data == nullptr) return false;

  const auto dshape = data->shape;
  CHECK_GE(dshape.size(), 2U)
      << "Pool2D only support input >= 2-D: input must have height and width";
  const auto param = attrs.as<MaxPool2dLocationAttrs>();
  CHECK(param != nullptr);

  Layout layout(param->layout);
  CHECK(layout.Contains(LayoutAxis::Get('H')) && layout.Contains(LayoutAxis::Get('W')) &&
        !layout.Contains(LayoutAxis::Get('h')) && !layout.Contains(LayoutAxis::Get('w')))
      << "Invalid layout " << layout << ". Pool2D layout must have H and W, which cannot be split";

  const auto hidx = layout.IndexOf(LayoutAxis::Get('H'));
  const auto widx = layout.IndexOf(LayoutAxis::Get('W'));

  IndexExpr pad_h, pad_w;
  if (param->padding.size() == 1) {
    pad_h = param->padding[0] * 2;
    pad_w = param->padding[0] * 2;
  } else if (param->padding.size() == 2) {
    // (top, left)
    pad_h = param->padding[0] * 2;
    pad_w = param->padding[1] * 2;
  } else if (param->padding.size() == 4) {
    // (top, left, bottom, right)
    pad_h = param->padding[0] + param->padding[2];
    pad_w = param->padding[1] + param->padding[3];
  } else {
    return false;
  }

  std::vector<IndexExpr> oshape(dshape.begin(), dshape.end());

  if (dshape[hidx].as<tir::AnyNode>()) {
    oshape[hidx] = dshape[hidx];
  } else {
    if (param->ceil_mode) {
      oshape[hidx] = ((dshape[hidx] + pad_h - param->pool_size[0] + param->strides[0] - 1) /
                      param->strides[0]) +
                     1;
    } else {
      oshape[hidx] = ((dshape[hidx] + pad_h - param->pool_size[0]) / param->strides[0]) + 1;
    }
  }
  if (dshape[widx].as<tir::AnyNode>()) {
    oshape[widx] = dshape[widx];
  } else {
    if (param->ceil_mode) {
      oshape[widx] = ((dshape[widx] + pad_w - param->pool_size[1] + param->strides[1] - 1) /
                      param->strides[1]) +
                     1;
    } else {
      oshape[widx] = ((dshape[widx] + pad_w - param->pool_size[1]) / param->strides[1]) + 1;
    }
  }

  // assign output type
  reporter->Assign(types[1], TensorType(oshape, data->dtype));
  return true;
}

Expr MakeMaxPool2dLocation(Expr data, Array<IndexExpr> pool_size, Array<IndexExpr> strides,
                           Array<IndexExpr> padding, std::string layout, bool ceil_mode) {
  auto attrs = make_object<MaxPool2dLocationAttrs>();
  attrs->pool_size = std::move(pool_size);
  attrs->strides = std::move(strides);
  attrs->padding = std::move(padding);
  attrs->ceil_mode = std::move(ceil_mode);
  attrs->layout = std::move(layout);
  static const Op& op = Op::Get("vision.max_pool2d_location");
  return Call(op, {data}, Attrs(attrs), {});
}

TVM_REGISTER_GLOBAL("relay.op.vision._make.max_pool2d_location")
    .set_body_typed(MakeMaxPool2dLocation);

RELAY_REGISTER_OP("vision.max_pool2d_location")
    .describe(R"doc(max_pool2d_location operator.

 - **data**: Input is 4D array of shape
             (batch_size, channels, height, width)
 )doc" TVM_ADD_FILELINE)
    .set_attrs_type<MaxPool2dLocationAttrs>()
    .set_num_inputs(1)
    .add_argument("data", "Tensor", "The input tensor.")
    .add_argument("pool_size", "Tensor", "The kernel size")
    .add_argument("strides", "Tensor", "The kernel stride size")
    .add_argument("padding", "Tensor", "The padding size")
    .add_argument("ceil_model", "Bool", "The ceil model")
    .set_support_level(5)
    .add_type_rel("MaxPool2dLocationRel", MaxPool2dLocationRel);

}  // namespace relay
}  // namespace tvm
